### PR TITLE
[CBR-467] fix STM and exception mistakes

### PR DESCRIPTION
**Pull request by avieth**
## Description

Like #3740 but with more fixes.

- The bracket_ order was wrong for walletWorker
- withWalletWorker is simpler and better: if either the reader or writer thread dies exceptionally, the other also dies. The writer would previously close the queue and then wait on the reader. Now this happens only on normal termination.
- `node` would deadlock in the call to `killNode` on exceptional shutdown. `killNode` would cancel the dispatcher thread, then close the network-transport endpoint. Since we used network-transport-tcp with the fair `QDisc`, which blocks on write, there would be deadlock, for the dispatcher thread would no longer be reading that `QDisc`. Thus `killNode` would throw a blocked indefinitely on `STM` transaction exception, overriding the `exitFailure (ExitCode 20)` raised by a `triggerShutdown`.

## Type of change

- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: wallet-new/src/Cardano/Wallet/Kernel/Actions.hs [hunk 1]
```diff
@@ -210,20 +213,13 @@ withWalletWorker wai k = do
...
-        -- Wait for the worker to finish, re-throwing any exceptions from it.
-        Async.wait as1)
-     (\_ -> k pushWA)
-
+  fmap snd $ withUnliftIO $ \(unlift :: UnliftIO m) -> Async.concurrently
+    -- Queue reader. If it dies, the writer dies too.
+    (walletWorker wai (STM.atomically getWA))
+    -- Queue writer. If it finishes, it closes the queue, causing the reader
+    -- to terminate normally. If it dies, forget about closing the queue, just
+    -- kill the reader.
```
Comment (avieth) [inline wallet-new/src/Cardano/Wallet/Kernel/Actions.hs]: `Async.concurrently` takes care of this for us. No need to explicitly `Async.link` the reader thread.

I want to emphasise that the meaning has changed here. If this `k` action dies exceptionally, then whatever is in the queue will not be read and `interp`'d.

Comment (avieth) [reply wallet-new/src/Cardano/Wallet/Kernel/Actions.hs]: (and I think that's a good thing)

File: wallet-new/src/Cardano/Wallet/Kernel/Actions.hs [hunk 2]
```diff
@@ -303,11 +298,15 @@ writeTMQueue (UnsafeTMQueue to tq) a = do
...
 -- If the 'TMQueue' is empty and closed, then this function returns 'Nothing'.
 -- Otherwise, if the 'TMQueue' is not closed, this function will block waiting
--- for new input.
+-- for new input, or for the queue to be closed.
 readTMQueue :: TMQueue a -> STM (Maybe a)
 readTMQueue (UnsafeTMQueue to tq) = do
-  STM.tryReadTQueue tq >>= \case
-     Just a -> pure (Just a)
-     Nothing -> STM.readTVar to >>= \case
-        TMQueueOpen -> Just <$> STM.readTQueue tq
```
[the remark this answers is not in the document]
Comment (avieth) [reply wallet-new/src/Cardano/Wallet/Kernel/Actions.hs]: Or, no, the `readTQueue` should `retry`, and then closing the queue should wake it up again, and everything should work...

**Comment by erikd** [comment]
I still don't trust the wallet integration tests. I would be tempted to `git commit --amend --no-edit --date=now` the last commit and force push to see if it happens again,

**Comment by erikd** [comment]
I still don't trust the wallet integration tests. I would be tempted to `git commit --amend --no-edit --date=now` the last commit and force push to see if it happens again,

**Pull request merged**